### DISCUSSION TITLE
detektProjectBaseline Fails with "Property 'formatting' is misspelled or does not exist" in a Modular Gradle Project

### DISCUSSION THREAD

**Discussion by VahidGarousi** [detekt/detekt#7839] [Q&A]
**Problem Description:**
We are working on a modular Android project with over 80 modules, and we want to incrementally enforce Detekt rules across the project. To begin, we started with BooleanPropertyNaming and disabled all other rules. Additionally, we aim to prevent committing changes that violate Detekt rules by checking only staged files during pre-commit hooks.

Here is the setup:
**AndroidDetektConventionPlugin** (applied to both AndroidApplicationConventionPlugin and AndroidLibraryConventionPlugin):
```kotlin
class AndroidDetektConventionPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.run {
            applyPlugins()
            detektGradle {
                configureDetekt(this)
            }
        }
    }

    private fun Project.applyPlugins() {
        pluginManager.apply {
            apply(libs.findLibrary("detekt-gradlePlugin").get().get().group.toString())
        }
    }
}


internal fun Project.configureDetekt(extension: DetektExtension) = extension.apply {
    config.setFrom("${rootDir}/config/detekt/detekt.yml")
    baseline = file("${rootDir}/config/detekt/detekt-baseline.xml")
    parallel = true
    ignoredFlavors = listOf("production", "staging")
    ignoredVariants =
        listOf("productionRelease", "productionDebug", "stagingRelease", "stagingDebug")
    tasks.withType<DetektCreateBaselineTask>().configureEach {
        jvmTarget = JavaVersion.VERSION_17.toString()
    }
    tasks.withType<Detekt>().configureEach {
        jvmTarget = JavaVersion.VERSION_17.toString()
        reports {
            xml.required.set(true)
            html.required.set(true)
            txt.required.set(false) // Disable TXT if not needed
            sarif.required.set(true)
            md.required.set(false)  // Disable Markdown if not needed
        }
    }
    dependencies {
        "detektPlugins"(libs.findLibrary("detekt-formatting").get())
    }
}
```

**2 - Root-level Task to Generate Baseline:**
```kotlin
val detektProjectBaseline by tasks.registering(DetektCreateBaselineTask::class) {
    description = "Overrides current baseline."
    ignoreFailures.set(true)
    parallel.set(true)
    setSource(files(rootDir))
    config.setFrom(files("$rootDir/config/detekt/detekt.yml"))
    baseline.set(file("$rootDir/config/detekt/baseline.xml"))
    include("**/*.kt")
    include("**/*.kts")
    exclude("**/resources/**")
    exclude("**/build/**")
}
```

**Issue:**
When I execute the following command to generate the baseline:
```bash
./gradlew detektProjectBaseline
```
I get the following error:
```
> Task :detektProjectBaseline FAILED
Property 'formatting' is misspelled or does not exist.
```

**What We've Tried:**
1 - Verified that detekt-formatting is included in the detektPlugins configuration in ```AndroidDetektConventionPlugin```

2 - Confirmed the content of detekt.
https://gist.github.com/VahidGarousi/67135e6e7bef85dae8a3738bd0f4740d

3 - Ensured that the Detekt tool version matches the Detekt Formatting plugin version:
```
toolVersion = "1.23.6"
```

**Our Requirements:**
We are starting with BooleanPropertyNaming as the only active rule and will gradually enable other rules.
In pre-commit hooks, we want Detekt to scan only staged files and prevent commits if any violations are detected.
Our structure and setup closely resemble the NowInAndroid project.

### **Update**

I successfully generated the **baseline file** using the task `detektProjectBaseline`. However, I noticed that there is no trace of the **`BooleanPropertyNaming`** rule in the generated baseline file.

I am confident that there are existing issues related to this rule in the project, but they are not being included in the baseline. Interestingly, the baseline file does include many other issues from across all modules.

Is there any reason why **`BooleanPropertyNaming`** might be skipped during the baseline generation? Could this be due to the allowed pattern in the configuration or something else?

Here’s my relevant configuration for reference:
```yaml
naming:
  active: true
  BooleanPropertyNaming:
    active: true
    allowedPattern: '^(is|has|are|can|was|should)'
```

Is there any step or configuration I’m missing to ensure BooleanPropertyNaming violations are added to the baseline? Any guidance would be greatly appreciated! 😊

**Comment by BraisGabin** [comment]
Multiple things:

> Additionally, we aim to prevent committing changes that violate Detekt rules by checking only staged files during pre-commit hooks.

I don't recommend you to do this. Check this code:

```kotlin
var foo = function() 
```

If `function` is defined in other file you could change that function from returning `Int` to return `Boolean`. You would not check the file that has that property because it was not changed. And now your codebase has a new issue. And, probably the worst thing, is that when someone else change that file with nothing related with that issue will find the issue and she will not be able to commit.

I highly recommend to use the default tasks of detekt for this reason. 

---

Did you check if the detekt task works as expected before you check if the baseline works as expected? The rule you choosed needs type resolution so I would bet that your issue is related with that.

**Reply by VahidGarousi** [reply]
A big thank you for your detailed and valuable response! Your insights were incredibly helpful, and I’ve tested everything you suggested. Fortunately, Gradle tasks like `detekt` and running Detekt on the entire project are working perfectly fine. I also have no issue running Detekt on the entire project.

### **The Main Challenge**:
I want to generate a **baseline** for the entire project that includes issues requiring **Type Resolution** (such as `BooleanPropertyNaming`). While Detekt works fine across the project, ensuring these type-resolution-related issues are included in the baseline is the part I want to set up correctly.

### **Additional Note**:
I can successfully generate a baseline file for each individual module, and it works without any problem. However, I don’t want to have a separate baseline file for each module. If I go this route, considering the 80 modules, along with flavors and build types, I would end up with **236 baseline files**, which is impractical to manage.

### **My Question**:
How can I configure Detekt to generate a **single baseline file** for the entire project while ensuring that issues requiring **Type Resolution** are included in this baseline? Any guidance or insights on this would be greatly appreciated. 😊

**Reply by BraisGabin** [reply]
OK, different points here too:

Probably you don't need to run detekt for all the flavours/build types on all the modules. Otherwise you will get a lot of duplicated issues and detekt will be way slower because it's doing duplicates work. In the project I work on I have a mixture between disable build variants that do nothing and then an alias task `detektAll` that runs the task that I care about for each module.

About creating a single baseline detekt doesn't provide anything out of the box. But the baseline file is a really simple xml. You can create a script to merge all the baseline in a single file. And then point to that file in the configuration.

I don't recommend this option. I prefer to have multiple baseline files and update it every day/week with a CI task. We don't allow to add new issues so this update just remove already fixed issues. Those fixes usually happen on refactors.

**Reply by VahidGarousi** [reply]
Could you please share a sample or a code snippet to better illustrate your approach? It would help clarify your suggestion and ensure we implement it as intended. 😊

**Reply by pacion** [reply]
figured out?
